lxc_ssh.py

```diff
@@ -222,69 +551,132 @@ def _add_args(self, b_command, b_args, explanation):
+        #
+
+        # sftp batch mode allows us to correctly catch failed transfers, but can
+        # be disabled if the client side doesn't support the option. However,
+        # sftp batch mode does not prompt for passwords so it must be disabled
+        # if not using controlpersist and using sshpass
+        if subsystem == 'sftp' and self.get_option('sftp_batch_mode'):
+            if conn_password:
+                b_args = [b'-o', b'BatchMode=no']
+                self._add_args(b_command, b_args, u'disable batch mode for sshpass')
+            b_command += [b'-b', b'-']
+
         if self._play_context.verbosity > 3:
             b_command.append(b'-vvv')
-        # Next, we add [ssh_connection]ssh_args from ansible.cfg.
-        #
-        if self._play_context.ssh_args:
+
+        # Next, we add ssh_args
+        ssh_args = self.get_option('ssh_args')
+        if ssh_args:
             b_args = [to_bytes(a, errors='surrogate_or_strict') for a in
-                    self._split_args(self._play_context.ssh_args)]
+                      self._split_ssh_args(ssh_args)]
             self._add_args(b_command, b_args, u"ansible.cfg set ssh_args")
 
-        # Now we add various arguments controlled by configuration file settings
-        # (e.g. host_key_checking) or inventory variables (ansible_ssh_port) or
-        # a combination thereof.
-        if not C.HOST_KEY_CHECKING:
+        # Now we add various arguments that have their own specific settings defined in docs above.
+        if not self.get_option('host_key_checking'):
             b_args = (b"-o", b"StrictHostKeyChecking=no")
             self._add_args(b_command, b_args, u"ANSIBLE_HOST_KEY_CHECKING/host_key_checking disabled")
-        if self._play_context.port is not None:
-            b_args = (b"-o", b"Port=" + to_bytes(self._play_context.port, nonstring='simplerepr', errors='surrogate_or_strict'))
+
+        self.port = self.get_option('port')
+        if self.port is not None:
+            b_args = (b"-o", b"Port=" + to_bytes(self.port, nonstring='simplerepr', errors='surrogate_or_strict'))
             self._add_args(b_command, b_args, u"ANSIBLE_REMOTE_PORT/remote_port/ansible_port set")
-        key = self._play_context.private_key_file
+
+        key = self.get_option('private_key_file')
         if key:
             b_args = (b"-o", b'IdentityFile="' + to_bytes(os.path.expanduser(key), errors='surrogate_or_strict') + b'"')
             self._add_args(b_command, b_args, u"ANSIBLE_PRIVATE_KEY_FILE/private_key_file/ansible_ssh_private_key_file set")
-        if not self._play_context.password:
+
+        if not conn_password:
             self._add_args(
                 b_command, (
                     b"-o", b"KbdInteractiveAuthentication=no",
-                    b"-o", b"PreferredAuthentications=publickey,gssapi-with-mic,gssapi-keyex,hostbased",
+                    b"-o", b"PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey",
```

Comment:
Minor question: why do you turn the authentication methods around?
In your PR "publickey" is now the last methog, and all the other ones need to be checked first, that is time-consuming.
Is there a logical explanation for this?

Reply:
Yeah, I blindly copied over the settings from ssh.py. I figured ssh.py had it in that order for a reason. Most likely this line change is not needed for the #30 fix.

Reply:
Ok, I think when you remove this from the PR I can merge it.

Reply:
My latest force push removes this change. 4e569a703a4ac5028f8b7bc9edf96187afe50bbf